flocker/ca/functional/test_validation.py

```diff
@@ -0,0 +1,274 @@
+                server_context_factory = peer_context_factory
+            else:
+                server_context_factory = validating_context_factory
+                client_context_factory = peer_context_factory
+
+            server_endpoint = SSL4ServerEndpoint(reactor, port,
+                                                 server_context_factory,
+                                                 interface='127.0.0.1')
+            d = server_endpoint.listen(
+                ServerFactory.forProtocol(SendingProtocol))
+            d.addCallback(lambda port: self.addCleanup(port.stopListening))
+            validating_endpoint = SSL4ClientEndpoint(
+                reactor, b"127.0.0.1", port, client_context_factory)
+            client_protocol = ReceivingProtocol()
+            result = connectProtocol(validating_endpoint, client_protocol)
+            result.addCallback(lambda _: client_protocol.result)
+            return result
+
+        def assert_does_not_validate(self, credential):
+            """
+            Asserts that a TLS handshake fails to happen between a client using
+            the validation logic and a server based on the given
+            credential.
+
+            :param FlockerCredential credential: The private key/certificate
+                to use for the server.
+
+            :return ``Deferred``: Fires on success (i.e. if no TLS handshake is
+                established).
+            """
+            return self.assertFailure(self.assert_validates(credential),
+                                      SSLError)
+
+        def test_same_ca_correct_type(self):
+            """
+            If the expected certificate type is generated by the same CA
+            then the validator will successfully validate it.
+            """
+            return self.assert_validates(
+                getattr(self.good_ca, good_certificate_name))
+
+        def test_different_ca_correct_type(self):
+            """
+            If the expected certificate type is generated by a different
+            CA then the validator will reject it.
+            """
+            return self.assert_does_not_validate(
+                getattr(self.another_ca, good_certificate_name))
+
+        def test_same_ca_wrong_type(self):
+            """
+            If the expected certificate is generated by the same CA but is of
+            the wrong type the validator will reject it.
+            """
+            return self.assert_does_not_validate(
+                getattr(self.another_ca, bad_name))
+
+        def test_same_ca_another_wrong_type(self):
+            """
+            If the expected certificate is generated by the same CA but is of
+            the wrong type the validator will reject it.
+
+            This is different wrong type than ``test_same_ca_wrong_type``.
+            """
+            return self.assert_does_not_validate(
+                getattr(self.another_ca, another_bad_name))
```

Comment:
Another case would be a certificate that purports to be signed by the correct CA but has a signature that does not validate.  This is probably an important case to have work so it would be nice to have a test - but at least it's OpenSSL's responsibility to make it work, not ours (so we'll have lots of company the day the CVE for this comes out).  I don't see an obvious way to construct such a certificate using the APIs in pyOpenSSL, so maybe it's harder than it's worth right now?  Or maybe it would be possible with a hack like getting a validly signed cert and then modifying it (eg change the subject which will result in the existing signature being invalidated)...

:question: